### PR TITLE
Allow mods and admins to delete messages in rooms

### PR DESCRIPTION
**Pull request by megamattron**
Add delete button to interface (placement is a beat weak atm). Handle deleted flag in initial state, reducers, message update handlers. Add server delete API endpoint.

This is a fix for #25.

@rickhanlonii - do you want to take a look at this before I merge it? It's my first end-to-end change through server - reducers - UI and so on.

### CODE REVIEW

**Comment by megamattron** [comment]
Should I just merge this @rickhanlonii ? For better or worse...

File: public/css/app.css [hunk 1]
```diff
@@ -359,3 +359,13 @@ li.chat-message-root .message-body, li.chat-message-short .message-body, li.chat
   cursor: pointer;
 }
 
+.visible-on-parent-hover {
+  display: none;
+  /*visibility: collapse;*/
```
Comment (rickhanlonii) [inline public/css/app.css]: Couple stray comments here

File: web/message/Message.jsx [hunk 2]
```diff
@@ -34,9 +35,23 @@ class Message extends Component {
   }
 
   renderHTMLMessage() {
-    let classes = 'message-body m-t-midxs';
-    const message = this.props.message.get('messageHtml').replace(/@([A-Za-z0-9_-]+)/g, this.renderFormattedMention);
-    return <div className={classes} dangerouslySetInnerHTML={ { __html: message } }></div>;
+    let classes = 'message-body m-t-midxs pos-rlt visible-on-hover-trigger';
```
Comment (rickhanlonii) [inline web/message/Message.jsx]: To clean this up to bit I would create three more render functions:

- `renderDeletedMessage`
- `renderMessageForMod`
- `renderMessage`

And have this method function only as a switch to call the correct render function.

File: web/redux/actions/chat-actions.js [hunk 3]
```diff
@@ -86,3 +87,14 @@ export function loadedMoreMessages(room, messages) {
 export function failedLoadingMoreMessages(error) {
   return { type: actions.CHAT_FAILED_LOADING_MESSAGES, error };
 }
+
+export function handleDeleteMessage(uuid) {
+  return (dispatch) => {
+    API.deleteMessage(uuid).then(() => {
+      console.log('Sent delete message request.');
```
Comment (rickhanlonii) [inline web/redux/actions/chat-actions.js]: Should probably remove the console statements. What I've done in the past is dispatch actions with no reducer handling (which would make them visible in the redux devtools but not to the end user).

**Comment by rickhanlonii** [comment]
@megamattron looks good, I left a few non-blocking comments

**Comment by megamattron** [comment]
Ran into a few problemos filtering out the deleted messages so opted instead to just render recent messages as deleted. They get fully filtered as they head into history. Gonna release this and see what goes wrong.

**Comment by megamattron** [comment]
BTW, something I did in here seems to have broken message grouping. It was attributing messages to other users so I rolled back the server to a previous version. Haven't looked into it in detail yet.

**Pull request merged**